File: world/grid_world.py
import random

class GridWorld:
    def __init__(self, width=10, height=10):
        self.width = width
        self.height = height
        self.npcs = []
        self.obstacles = set()

        self.tiles = [["." for _ in range(self.width)] for _ in range(self.height)]

    def generate_obstacles(self, count):
        attempts = 0
        while len(self.obstacles) < count and attempts < count * 10:
            x = random.randint(0, self.width - 1)
            y = random.randint(0, self.height - 1)

            if (x, y) not in self.obstacles:
                if all(npc.x != x or npc.y != y for npc in self.npcs):
                    self.add_obstacle(x, y)

            attempts += 1

    def add_npc(self, npc):
        npc.world = self
        self.npcs.append(npc)

    def update(self):
        for npc in self.npcs:
            npc.update()

    def render(self):
        grid = [row[:] for row in self.tiles]

        for (x, y) in self.obstacles:
            if self.is_within_bounds(x, y):
                grid[y][x] = "#"

        for npc in self.npcs:
            if self.is_within_bounds(npc.x, npc.y):
                grid[npc.y][npc.x] = npc.name[0].upper()

        for row in grid:
            print(' '.join(row))
        print()

    def is_within_bounds(self, x, y):
        return 0 <= x < self.width and 0 <= y < self.height
    
    def add_obstacle(self, x, y):
        if self.is_within_bounds(x, y):
            self.obstacles.add((x, y))

    def is_obstacle(self, x, y):
        return (x, y) in self.obstacles

    def is_walkable(self, x, y):
        return self.is_within_bounds(x, y) and not self.is_obstacle(x, y)

File: C_SHARP/GridWorld.cs
using OpenNPC_CSharp_ver;
using System;
using System.Collections.Generic;

namespace OpenNPC_CSharp_ver
{
	public class GridWorld
	{
		public int Width { get; }
		public int Height { get; }
		public HashSet<(int, int)> Obstacles { get; }
		public List<NPC> NPCs { get; }

		public GridWorld(int width = 25, int height = 25)
		{
			Width = width;
			Height = height;
			Obstacles = new HashSet<(int, int)>();
			NPCs = new List<NPC>();
		}

		public void AddNPC(NPC npc)
		{
			npc.World = this;
			NPCs.Add(npc);
		}

		public bool IsWithinBounds(int x, int y)
		{
			return x >= 0 && x < Width && y >= 0 && y < Height;
		}

		public bool IsObstacle(int x, int y)
		{
			return Obstacles.Contains((x, y));
		}

		public bool IsWalkable(int x, int y)
		{
			return IsWithinBounds(x, y) && !IsObstacle(x, y);
		}

		public void GenerateObstacles(int count)
		{
			var rnd = new Random();
			int attempts = 0;
            while (Obstacles.Count < count && attempts < count * 10)
            {
				int x = rnd.Next(0, Width);
				int y = rnd.Next(0, Height);

				if (!Obstacles.Contains((x, y)) && NPCs.TrueForAll(npc => npc.X != x || npc.Y != y))
				{
					AddObstacle(x, y);
				}
				++attempts;
            }
        }

		public void AddObstacle(int x, int y)
		{
			if (IsWithinBounds(x, y))
			{
				Obstacles.Add((x, y));
			}
		}

		public void Render()
		{
			for (int y = 0; y < Height; y++)
			{
				for (int x = 0; x < Width; x++)
				{
					bool hasNpc = false;
					foreach (var npc in NPCs)
					{
						if (npc.X == x && npc.Y == y)
						{
							Console.Write(npc.Name[0]);
							hasNpc = true;
							break;
						}
					}
					if (!hasNpc)
					{
						Console.Write(IsObstacle(x, y) ? "#" : ".");
					}
				}
				Console.WriteLine();
			}
		}

		public void Update()
		{
			foreach (var npc in NPCs)
			{
				npc.Update();
			}
		}

	}
}